keps/sig-testing/20190605-presubmitconfig-in-tested-repo.md

```diff
@@ -0,0 +1,259 @@
+* [Risks and Mitigations](#risks-and-mitigations)
+   * [Plugins may have an inconsitent view on what presubmits exist for a repo](#plugins-may-have-an-inconsitent-view-on-what-presubmits-exist-for-a-repo)
+   * [Someone could accidentally delete all presubmits in prow.yaml and have their PR instantly merged](#someone-could-accidentally-delete-all-presubmits-in-prowyaml-and-have-their-pr-instantly-merged)
+* [Implementation History](#implementation-history)
+
+## Release Signoff Checklist
+
+**ACTION REQUIRED:** In order to merge code into a release, there must be an issue in [kubernetes/enhancements] referencing this KEP and targeting a release milestone **before [Enhancement Freeze](https://github.com/kubernetes/sig-release/tree/master/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core Kubernetes i.e., [kubernetes/kubernetes], we require the following Release Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This document proposes to change the Prow presubmit handling to optionally version some or
+all presubmits in the same repository that also contains the code that is being tested.
+
+## Motivation
+
+Currently, all jobs for Prow are configured in a central `infra-config` repository that is
```

Comment:
Needs to clearly note that the reason for this is a decision on trust -- deployments that are fully OSS like for prow.k8s.io cannot trust the users that are checking in job config. 

Having jobs in a repo allows a PR to get itself to merge by removing all the blocking presubmits.

Reply:
Can you elaborate what attack vector this opens that doesn't exist today?

If I can get my pull to be tested on k/k today, I can change the scripts that are executed by the job to get access to credentials.

The differences I can think off is that additional presets can be added and that its possible to set the cluster on the job.

I will add a section about this in the `Risks and mitigation` section.

Reply:
> Having jobs in a repo allows a PR to get itself to merge by removing all the blocking presubmits.

As noted in the `risks and mitigation` section, it is expected that PRs need a review so the author can not just get it merged without having someone look at it. Also, Prow doesn't even test PRs of untrusted ppl

Reply:
IMHO this doesn't have too much to do with OSS or not. We will definitely activate this on our OSS repos, the fact that external contributions won't get tested without a member using `/test all` is enough safety here